src/blockchain.rs

```diff
@@ -39,20 +58,169 @@ impl Blockchain {
 
     /// Returns the latest `Block` as part of the longest chain
     pub fn get_latest_block_index(&self) -> Option<&BlockIndex> {
-        self.index.blocks.last()
+        match self.index.last_block_hash {
+            Some(last_block_hash) => self.index.block_map.get(&last_block_hash),
+            None => None,
+        }
+    }
+
+    pub fn get_genesis_block_index(&self) -> Option<&BlockIndex> {
+        match self.index.genesis_block_hash {
+            Some(genesis_block_hash) => self.index.block_map.get(&genesis_block_hash),
+            None => None,
+        }
     }
 
     /// Append `Block` to the index of `Blockchain`
     ///
     /// * `block` - `Block` appended to index
-    pub fn add_block(&mut self, block: Block) {
+    pub fn try_add_block(&mut self, block: Block) -> crate::Result<()> {
+        // TODO -- we want to lock writes to the blockchain while we're adding a block
+        // To do that, we should implement a RwLock around the `Blockchain` index to verify
+        // that there's only one thread that is writing to the chain at once time
```

Comment:
I don't think anyone should read from the blockchain during this time either because of things that look similar to the "ABA problem".

Put another way, the state of blockchain is inconsistent during add_block and I think even reading from it could be dangerous...